include/secp256k1_musig.h

```diff
@@ -0,0 +1,579 @@
+ *  MuSig, musig and MuSig2 synonymously unless noted otherwise.
+ */
+
+/** Opaque data structures
+ *
+ *  The exact representation of data inside is implementation defined and not
+ *  guaranteed to be portable between different platforms or versions. If you
+ *  need to convert to a format suitable for storage, transmission, or
+ *  comparison, use the corresponding serialization and parsing functions.
+ */
+
+/** Opaque data structure that caches information about public key aggregation.
+ *
+ *  Guaranteed to be 197 bytes in size. It can be safely copied/moved. No
+ *  serialization and parsing functions (yet).
+ */
+typedef struct {
+    unsigned char data[197];
+} secp256k1_musig_keyagg_cache;
+
+/** Opaque data structure that holds a signer's _secret_ nonce.
+ *
+ *  Guaranteed to be 132 bytes in size.
+ *
+ *  WARNING: This structure MUST NOT be copied or read or written to directly. A
+ *  signer who is online throughout the whole process and can keep this
+ *  structure in memory can use the provided API functions for a safe standard
+ *  workflow.
+ *
+ *  Copying this data structure can result in nonce reuse which will leak the
+ *  secret signing key.
+ */
+typedef struct {
+    unsigned char data[132];
+} secp256k1_musig_secnonce;
+
+/** Opaque data structure that holds a signer's public nonce.
+ *
+ *  Guaranteed to be 132 bytes in size. It can be safely copied/moved. Serialized
+ *  and parsed with `musig_pubnonce_serialize` and `musig_pubnonce_parse`.
+ */
+typedef struct {
+    unsigned char data[132];
+} secp256k1_musig_pubnonce;
+
+/** Opaque data structure that holds an aggregate public nonce.
+ *
+ *  Guaranteed to be 132 bytes in size. It can be safely copied/moved.
+ *  Serialized and parsed with `musig_aggnonce_serialize` and
+ *  `musig_aggnonce_parse`.
+ */
+typedef struct {
+    unsigned char data[132];
+} secp256k1_musig_aggnonce;
+
+/** Opaque data structure that holds a MuSig session.
+ *
+ *  This structure is not required to be kept secret for the signing protocol to
+ *  be secure. Guaranteed to be 133 bytes in size. It can be safely
+ *  copied/moved. No serialization and parsing functions (yet).
+ */
+typedef struct {
+    unsigned char data[133];
+} secp256k1_musig_session;
+
+/** Opaque data structure that holds a partial MuSig signature.
+ *
+ *  Guaranteed to be 36 bytes in size. Serialized and parsed with
+ *  `musig_partial_sig_serialize` and `musig_partial_sig_parse`.
+ */
+typedef struct {
+    unsigned char data[36];
+} secp256k1_musig_partial_sig;
+
+/** Parse a signer's public nonce.
+ *
+ *  Returns: 1 when the nonce could be parsed, 0 otherwise.
+ *  Args:    ctx: pointer to a context object
+ *  Out:   nonce: pointer to a nonce object
+ *  In:     in66: pointer to the 66-byte nonce to be parsed
+ */
+SECP256K1_API int secp256k1_musig_pubnonce_parse(
+    const secp256k1_context *ctx,
+    secp256k1_musig_pubnonce *nonce,
+    const unsigned char *in66
```

Comment:
Wouldn't it be better to declare this as `const unsigned char (*in66)[66]`? Aside from clarity it could help static analysis tools or perhaps bindings generators to generate correct bindings.

Reply:
That's an interesting point. Right now, we consistently use declarations of the form `const unsigned char *in66` consistently in the library. I prefer to have a standard across the whole API and would avoid just changing this in the musig module.

Reply:
That makes sense. Might be nice to change all of them at some point unless there's a problem I'm missing.

Reply:
For instance in Rust we have [`bindgen`](https://crates.io/crates/bindgen), a tool that generates bindings automatically. I've tried to pass it this header file:

```c
void foo(const unsigned char *arg);
void bar(const unsigned char arg[32]);
void baz(const unsigned char (*arg)[32]);
```

And it generated this code:

```rust
extern "C" {
    pub fn foo(arg: *const ::std::os::raw::c_uchar);
}
extern "C" {
    pub fn bar(arg: *const ::std::os::raw::c_uchar);
}
extern "C" {
    pub fn baz(arg: *const [::std::os::raw::c_uchar; 32usize]);
}
```

As you can see, the last one has a different type. And while it's the same from ABI perspective, there's a real difference between them in Rust since you can directly pass in `&array` without any casts because the types match making the code more obviously correct.

Currently, we aren't using `bidgen` in rust-secp256k1 but we might in the future and I'd be surprised if similar tools don't exist in other languages.

Also I'm not familiar with the current state of static analysis in C but I think a reasonable static analyzer should be able to see that statements like `out[33] = 42` or `baz(&twentyone_item_array)` are obviously wrong. `gcc` already complains about the these when compiling with `-O2 -W -Wall -Warray-bounds`.

Reply:
Oh, I found that `bindgen` has an option to generate `*const [c_uchar; 32]` instead of `*const c_uchar` for arguments defined as `const unsigned char arg[32]`, so maybe even that is already valuable? (IIUC not API-breaking) I think it has documentation value too.

Reply:
@Kixunil We discussed this in our IRC meeting today. Can you try if bindgen understands `arg[static 32]`? This is a special syntax in C99, see https://hamberg.no/erlend/posts/2013-02-18-static-array-indices.html